### PR TITLE
Fixes #28740 - Use ForemanModal as a confirmation dialog

### PR DESCRIPTION
**Pull request by xprazak2**

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitBtn.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitBtn.js]: From a UI perspective, people don't read the text of dialog boxes; they just read the text of the button and decide to click based on that.  So, the button label should be the action the user is about to do, such as "Delete", etc.  I think that hard-coding it to "Submit" is too all-purpose.  Can I pass the button text as a prop instead of having it set in stone as 'Submit' ?  Also, for destructive things like deleting, I would want to be able to set `bsStyle` to `warning` or `danger`.  Should pass that down too.

(I do really like the use of `simpleLoader` btw)

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js]: I'm not sucking, I'm experiencing success.  Prefer the unabbreviated names `onSuccess` and `onError` :smiley:

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js]: These dispatch `SET_MODAL_START_SUBMITTING` and `SET_MODAL_STOP_SUBMITTING` actions bound to a particular modal ID.  But I think it makes it confusing to pass in ID-bound functions here, in an action creator.  I would prefer instead if you could accept the modal ID as an argument and pass it through.

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js:44]: Might think about incorporating @laviro #7261 here once merged

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalActions.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalActions.js:23]: Nice elegant way to expand the checking for existence of a modal to multiple actions.  I was actually thinking that throwing errors for this is too annoying for developers anyway, and that the fact that we use constants for modal ID's is enough to prevent modal ID confusion.  I may put in a PR soon to remove the error throwing (just FYI).  thoughts welcome

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.test.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.test.js]: This is exactly the way I envisioned adding functionality to ForemanModals. :+1: :+1:

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.js]: If `submitProps` are supplied then maybe it should render a `<ForemanModal.Footer />` for you, as it does here with the Header.  Wouldn't want to swallow submitProps because I forgot to put in the `<ForemanModal.Footer />` child.

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalActions.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalActions.js:45]: Thanks for adding these here

**Review by jeremylenz — COMMENTED** [review]
Nice work @xprazak2!  Comments below

**Comment by sharvit** [comment]
Need a rebase, It will force you to use the new CSF format for your stories, see: https://github.com/theforeman/foreman-js/tree/master/packages/stories#component-story-format-csf

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalUsage.stories.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalUsage.stories.js]: Can `<ForemanModal.Footer />` be omitted now?

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalUsage.stories.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalUsage.stories.js]: The `story: { name: ... }` object is not required if you don't need to change the name or have special characters.  Storybook will transform `asConfirmationDialogWithCustomButtons` to "As confirmation dialog with custom buttons".  That said, if you want to change that word from "custom" to "customized" you could leave it as is, or change the function name.

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancel.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancel.js]: Prefer doing this with CSS (`margin-left` on the rightmost button should do it.)

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js]: The name of `onError` makes me think this is an error handler function that will automatically be called when an error is thrown, but instead it seems like it's just retrieving the error message to be displayed in the Toast.  Kind of like a selector, but instead of selecting from Redux state it's selecting from the API response.  Maybe rename to `getErrorMsg` or something similar?

**Review by jeremylenz — COMMENTED** [review]
Didn't retest functionality yet (waiting on #6818 update) but left some code comments below.

**Comment by xprazak2** [comment]
I've updated both this and how it is used on hw models page.

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js]: nit: maybe we could change `selectModalStateById` to return empty object `{ }` if doesn't exist,
than we won't need to check if it exist or not in each selector, e.g:
`export const selectIsModalSubmitting = (state, id) => selectModalStateById(state, id).submitting`

imo for selectors which are constantly being called it could improve performance.

File: webpack/assets/javascripts/react_app/components/ForemanModal/__tests__/__snapshots__/ForemanModal.test.js.snap
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ForemanModal/__tests__/__snapshots__/ForemanModal.test.js.snap:9]: we have some places where we use `submitting` and others where we use `isSubmitting`
if they both have the same meaning should we use the same name? I like the `is__` convention for Boolean values.

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalUsage.stories.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalUsage.stories.js]: can you update this as well?

File: webpack/assets/javascripts/react_app/components/ForemanModal/index.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: would it be a better performance to declare this function outside the useEffect, like the `onClose` function?

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/index.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/index.js]: I wonder if we should move the `SubmitOrCancel.js` code to this file, instead of passing props to both submit and cancel even if usually we have only one main component in `index.js`, e.g: 
```js
  <React.Fragment>
    <SubmitBtn
      onSubmit={onSubmit}
      submitting={submitting}
      {...submitBtnProps}
    />
    <CancelBtn onCancel={onCancel} disabled={submitting} {...cancelBtnProps} />
  </React.Fragment>
```

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/SubmitOrCancelActions.js]: I believe you can destruct the error object into something like: 
`const { response: { status, data: { error: { message, full_messages } } } = {} } = error`

and it will make the code below look much cleaner.

**Review by Ron-Lavi — COMMENTED** [review]
Thanks @xprazak2 ! looks great :)
Left some inline comments

File: webpack/assets/javascripts/react_app/components/ForemanModal/index.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: @laviro I don't think this is necessary, see [this part](https://reactjs.org/docs/hooks-faq.html#are-hooks-slow-because-of-creating-functions-in-render) of the docs where React weighs in on it.  They do provide some other hooks to increase performance, but I think any performance optimization effort isn't really worth it unless performance is measured first.

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/index.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/SubmitOrCancel/index.js]: SubmitOrCancel/index.js exports the 'connected' component (using redux hooks though, so 'hooked component'?), we would need to wrap it in `<Provider>` for tests.

**Comment by xprazak2** [comment]
Updated, main changes are renaming states for modal
* submitting -> isSubmitting
* open -> isOpen

and selector for modal state now returns empty object if modal is not registered.

File: webpack/assets/javascripts/react_app/components/ForemanModal/index.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: With the change to the `selectModalStateById` selector, since `{}` is truthy it might cause a problem here?

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: I'm thinking it might be useful to just add a new selector `selectDoesForemanModalExist` and use that instead

Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: Updated, this would have caused problems. Nice catch!

Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: I added a helper, selector would work as well. Let me know which one you prefer.

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: @xprazak2 I would actually prefer the selector, I think it reads more clearly that way.  thanks for addressing it 😄

**Comment by xprazak2** [comment]
Updated, I changed helper to selector.

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js]: The selector should be named `selectModalExists` just to conform to the convention.  For readability elsewhere I'm fine with the extra line of code to assign its return value to a variable: `const modalExists = selectModalExists(...)`

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js:25]: I'm thinking maybe `id` should be a required prop, to avoid undefined values being passed on.

**Review by jeremylenz — COMMENTED** [review]
Thanks @xprazak2!  Changes since last review look good, just a couple comments

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js:25]: It gets pulled from modal context in `ForemanModalFooter` and `SubmitOrCancel` has it as a required prop.

Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js:25]: @xprazak2 sorry, I think I commented in the wrong place.  I meant the ID passed to the delete modal to tell it which model to delete.

Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js:25]: I see. The problem is that until the delete button is clicked, we do not know id of a model we want to delete, so we would get a failed prop type...

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.js]: nitpick: for keeping code readabilty, how about this
```js 
footerToRender = footerChild || <ForemanModalFooter />
// or
const footerToRender = footerChild || renderDefaultFooter(submitProps)
// same for header
const headerToRender = headerChild || renderDefaultHeader(title)
```

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js]: why creating an Immutable empty object?  selectors don't affect redux's store, just read from it.

File: webpack/assets/javascripts/react_app/components/ForemanModal/index.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: if id has been changed, also `modalExists` changes,  is it needed to add `id` here?

File: webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/ForemanModal/subcomponents/ForemanModalFooter.js]: Needs a translation ?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @xprazak2! awesome feature :)
Just few nitpicks

File: webpack/assets/javascripts/react_app/components/ForemanModal/index.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/components/ForemanModal/index.js]: I agree `id` doesn't need to be added.  Also I feel we should stop disabling the eslint rule here.

File: webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js]: So I do not get `undefined` state when checking if modal is open:

undefined-open

Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js]: I mean using `Immutable` here is overkill,  an empty object will do the job :)
```js
state.foremanModals && (state.foremanModals[id] || {})
```

Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalSelectors.js]: Updated.

**Review by amirfefer — CHANGES_REQUESTED** [review]
@xprazak2  looks good! 
just few snapshots are broken

**Comment by amirfefer** [comment]
[test katello]

**Review by amirfefer — APPROVED** [review]
Thanks @xprazak2  ! 
LGTM :+1:

**Comment by sharvit** [comment]
Waiting for a green light from @laviro and @jeremylenz before merging. Thanks!

**Comment by Ron-Lavi** [comment]
@sharvit @xprazak2 Thanks, my concerns were addressed :) LGTM

**Comment by jeremylenz** [comment]
@sharvit LGTM :+1:

**Comment by sharvit** [comment]
Thanks @xprazak2 @jeremylenz @laviro @amirfefer :rocket:

**Pull request merged**
